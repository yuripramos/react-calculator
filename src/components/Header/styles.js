import styled, { css } from "styled-components";
import { rem, media } from "../../styles/tools";
import { white, grey40 } from "../../styles/settings";

export const HeaderWrapper = styled.header`
  height: ${rem(88)};
  padding-top: ${rem(10)};
  background: ${grey40};
  text-align: center;
  ${media.md(css`
    text-align: center;
  `)};
  ${media.lg(css`
    height: ${rem(110)};
    padding-top: ${rem(6)};
  `)};
`;

export const Title = styled.h1`
  font-family: "OpenSans Regular";
  font-size: ${rem(25)};
  text-align: center;
  color: ${white};
  letter-spacing: 1.5,
  text-shadow: 0.1875rem 0.1rem rgba(132, 129, 129, 0.15);
`;

export const TitleWrapper = styled.div`
  display: inline-flex;
`;

export const SubTitle = styled.h2`
  font-family: "OpenSans Regular";
  font-size: ${rem(10)};
  text-align: center;
  color: ${white};
  text-shadow: ${rem(3)} ${rem(3)} rgba(0, 0, 0, 0.25);
  ${media.md(css`
    font-size: ${rem(24)};
    text-align: left;
  `)};
`;

export const InputWrapper = styled.div`
  display: inline-flex;
  align-items: center;
  justify-content: center;
  width: 100%;
`;

export const ListIcons = styled.div`
  display: flex;
  width: ${rem(200)};
  margin-top: ${rem(15)};
`;

export const FeaturedImage = styled.div`
  width: 100%;
  height: 100%;
  margin: 0;
`;

export const ImgHeader = styled.img`
  width: ${rem(60)};
  height: ${rem(60)};
  ${media.lg(css`
    width: ${rem(80)};
    height: ${rem(80)};
  `)};
`;
