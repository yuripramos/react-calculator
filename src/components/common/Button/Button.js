import React from "react";
import { bool, node, string, func, number } from "prop-types";

import { ButtonTag, RouterLinkTag } from "./styles";
import LocalLoading from "../../common/LocalLoading";
function Button({
  children,
  onClick,
  type,
  disabled,
  width,
  opacity,
  to,
  isCallToAction,
  loading
}) {
  if (to) {
    return (
      <RouterLinkTag
        width={width}
        onClick={onClick}
        isCallToAction={isCallToAction}
        to={to}
        disabled={disabled}
        loading={loading.toString()}
      >
        {loading ? <LocalLoading /> : children}
      </RouterLinkTag>
    );
  }
  return (
    <ButtonTag
      width={width}
      type={type}
      disabled={disabled}
      onClick={onClick}
      opacity={opacity}
      loading={loading}
    >
      {children}
    </ButtonTag>
  );
}

Button.displayName = "Button";

Button.defaultProps = {
  type: "button",
  disabled: false,
  width: null,
  onClick: null,
  className: null,
  to: null,
  loading: false,
};

Button.propTypes = {
  children: node.isRequired, // eslint-disable-line
  isCallToAction: bool,
  disabled: bool,
  type: string,
  width: string,
  onClick: func,
  opacity: number,
  loading: bool
};

export default Button;
