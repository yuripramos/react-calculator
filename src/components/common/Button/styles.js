import React from "react";
import styled, { css } from "styled-components";
import { rem, media } from "../../../styles/tools";
import { Link as RouterLink } from "react-router-dom";
import { white, green, green20, black30 } from "../../../styles/settings";

export const ButtonTag = styled.button`
  font-family: "RobotoSlab Regular";
  border-radius: ${rem(4)};
  border: 0;
  cursor: pointer;
  display: inline-block;
  font-size: ${rem(12)};
  letter-spacing: ${rem(0.5)};
  line-height: ${rem(45)};
  outline: none;
  padding: ${rem(5)};
  position: relative;
  text-align: center;
  text-decoration: none;
  color: ${white};
  transition: 0.2s ease-in-out color, 0.2s ease-in-out background,
    0.2s ease-in-out box-shadow, 0.2s ease-in-out border, 0.1s transform;
  white-space: nowrap;
  width: ${props => (props.width ? rem(props.width) : "100%")};
  user-select: none;
  opacity: ${props => (props.opacity && props.disabled ? props.opacity : 1)};
  background: transparent;
  transform: translateY(15px);

  ${media.md(css`
    width: ${props => (props.width ? rem(props.width) : "auto")};
  `)};

  ${media.lg(css`
    font-size: ${rem(13)};
    line-height: ${rem(48)};
  `)};

  :active {
    transform: translateY(${rem(1)});
  }
  :hover {
    background: ${white};
    color: ${black30};
    transform: translateY(20px);
  }

  ${({ isCallToAction }) =>
    isCallToAction &&
    css`
      background-color: ${green};
      box-shadow: 0 ${rem(2)} ${rem(7)} ${rem(2)} rgba(100, 100, 100, 0.3);
      color: ${white};
      :hover {
        background: ${green20};
      }
    `};
`;

export const RouterLinkTag = ButtonTag.withComponent(
  ({ isCallToAction, ...props }) => <RouterLink {...props} /> // eslint-disable-line
);
