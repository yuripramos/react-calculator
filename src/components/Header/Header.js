import React, { Component } from "react";
import { shape, func, string, bool } from "prop-types";
import { MODAL_TYPES } from "../common/Modal/Modal";

import { HeaderWrapper, Title, TitleWrapper, InputWrapper } from "./styles";
import { Container, Row, Column } from "../../styles/grid";
import Button from "../common/Button";
import Input from "../common/Input";

class Header extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isLoading: false
    };
    this.handleConfirmButton = this.handleConfirmButton.bind(this);
    this._isMounted = false;
  }

  async handleConfirmButton() {
    this.setState({
      isLoading: true
    });
    await this.props.forwardGeocode(this.props.search.city);
    if(this._isMounted) {
      this.setState({
        isLoading: false
      });
      this.props.clearSearch();
    }
  }
  componentDidMount() {
    this._isMounted = true;
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  render() {
    const {
      handleUserInput,
      search,
      openModal,
      closeModal,
      isError,
      clearError
    } = this.props;
    const { isLoading } = this.state;

    if (isError) {
      setTimeout(() => {
        openModal({
          title: "We are sorry but we were unable to service your request",
          type: MODAL_TYPES.DEFAULT,
          icon: "attention",
          description: ["Try be more more specific e.g. Brussels, Belgium", "If the error persist try again in 5 minutes"],
          onClose: () => {
            clearError();
            return closeModal()
          }
        });
      },300);
    }

    return (
      <HeaderWrapper>
        <Container>
          <Row>
            <Column>
              <TitleWrapper>
                <Title>Forecast App</Title>
              </TitleWrapper>
            </Column>
          </Row>
          <Row>
            <Column>
              <InputWrapper>
                <Input
                  onChange={handleUserInput}
                  value={search.city}
                  placeholder={
                    "city, country to forecast =)"
                  }
                  height={40}
                  width={100}
                  type="text"
                  name="city"
                />
                <Button
                  width={"80"}
                  to={{
                    pathname: "/forecast",
                    search: `?search=${search.city}`
                  }}
                  onClick={this.handleConfirmButton}
                  loading={isLoading}
                >
                  SEARCH
                </Button>
              </InputWrapper>
            </Column>
          </Row>
        </Container>
      </HeaderWrapper>
    );
  }
}

export default Header;

Header.defaultProps = {
  search: {
    city: ""
  }
};

Header.propTypes = {
  handleUserInput: func,
  clearSearch: func,
  clearError: func,
  forwardGeocode: func,
  isError: bool,
  search: shape({
    city: string
  }),
  openModal: func,
  closeModal: func
};
