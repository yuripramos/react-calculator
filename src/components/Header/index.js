import { connect } from "redux-zero/react";
import { combineActions } from "redux-zero/utils";

import Header from "./Header";
import Forecastactions from "../../actions/Forecast";
import modalActions from "../../actions/modal";

export default connect(
  ({
    handleUserInput,
    search,
    clearSearch,
    isLoading,
    isError,
    clearError
  }) => ({
    handleUserInput,
    search,
    clearSearch,
    isLoading,
    isError,
    clearError
  }),
  combineActions(Forecastactions, modalActions)
)(Header);
