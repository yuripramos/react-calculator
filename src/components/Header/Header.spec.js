import React from "react";
import { shallow } from "enzyme";
import Header from "./Header";
import { spy } from "sinon";
import Button from "../common/Button";

const props ={
  handleUserInput: jest.fn(),
  search: {
    city: "brussels"
  },
  clearSearch: jest.fn(),
  isLoading: false,
  isError: false

}
describe("Header component", () => {
  it("should match snapshot with props", () => {
    expect(shallow(<Header {...props}/>)).toMatchSnapshot();
  });
  it("should call handleConfirmButton", () => {
    spy(Header.prototype, "handleConfirmButton");

    const wrapper = shallow(<Header {...props} />);
    wrapper.find(Button).simulate("click");

    expect(Header.prototype.handleConfirmButton.calledOnce);

    Header.prototype.handleConfirmButton.restore();
  });

});
