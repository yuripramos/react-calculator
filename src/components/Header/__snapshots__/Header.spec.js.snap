// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should match snapshot with props 1`] = `
<styled.header>
  <Container>
    <Row>
      <Column>
        <styled.div>
          <styled.h1>
            Forecast App
          </styled.h1>
        </styled.div>
      </Column>
    </Row>
    <Row>
      <Column>
        <styled.div>
          <Input
            checked={false}
            height={40}
            name="city"
            onChange={[MockFunction]}
            placeholder="city, country to forecast =)"
            type="text"
            value="brussels"
            width={100}
          />
          <Button
            className={null}
            disabled={false}
            loading={false}
            onClick={[Function]}
            to={
              Object {
                "pathname": "/forecast",
                "search": "?search=brussels",
              }
            }
            type="button"
            width="80"
          >
            SEARCH
          </Button>
        </styled.div>
      </Column>
    </Row>
  </Container>
</styled.header>
`;
